builtin/logical/mongodb/path_roles.go

```diff
@@ -0,0 +1,182 @@
+			},
+			"roles": &framework.FieldSchema{
+				Type:        framework.TypeString,
+				Description: "MongoDB roles to assign to the users generated for this role.",
+			},
+		},
+
+		Callbacks: map[logical.Operation]framework.OperationFunc{
+			logical.ReadOperation:   b.pathRoleRead,
+			logical.UpdateOperation: b.pathRoleCreate,
+			logical.DeleteOperation: b.pathRoleDelete,
+		},
+
+		HelpSynopsis:    pathRoleHelpSyn,
+		HelpDescription: pathRoleHelpDesc,
+	}
+}
+
+func (b *backend) Role(s logical.Storage, n string) (*roleEntry, error) {
+	entry, err := s.Get("role/" + n)
+	if err != nil {
+		return nil, err
+	}
+	if entry == nil {
+		return nil, nil
+	}
+
+	var result roleEntry
+	if err := entry.DecodeJSON(&result); err != nil {
+		return nil, err
+	}
+
+	return &result, nil
+}
+
+func (b *backend) pathRoleDelete(
+	req *logical.Request, data *framework.FieldData) (*logical.Response, error) {
+	err := req.Storage.Delete("role/" + data.Get("name").(string))
+	if err != nil {
+		return nil, err
+	}
+
+	return nil, nil
+}
+
+func (b *backend) pathRoleRead(
+	req *logical.Request, data *framework.FieldData) (*logical.Response, error) {
+	role, err := b.Role(req.Storage, data.Get("name").(string))
+	if err != nil {
+		return nil, err
+	}
+	if role == nil {
+		return nil, nil
+	}
+
+	rolesJsonBytes, err := json.Marshal(role.MongoDBRoles)
+	if err != nil {
+		return nil, err
+	}
+
+	return &logical.Response{
+		Data: map[string]interface{}{
+			"db": role.DB,
+			"roles": string(rolesJsonBytes),
+		},
+	}, nil
+}
+
+func (b *backend) pathRoleList(
+	req *logical.Request, d *framework.FieldData) (*logical.Response, error) {
+	entries, err := req.Storage.List("role/")
+	if err != nil {
+		return nil, err
+	}
+
+	return logical.ListResponse(entries), nil
+}
+
+func (b *backend) pathRoleCreate(
+	req *logical.Request, data *framework.FieldData) (*logical.Response, error) {
+	name := data.Get("name").(string)
+	roleDB := data.Get("db").(string)
+
+	// Example roles JSON: [ "readWrite", { "role": "readWrite", "db": "test" } ]
+	var roles []interface{}
+	rolesJson := []byte(data.Get("roles").(string))
+	if len(rolesJson) > 0 {
+		var rolesArray []interface{}
+		err := json.Unmarshal(rolesJson, &rolesArray)
+		if err != nil {
+			return nil, err
+		}
+		for _, rawItem := range rolesArray {
+			switch item := rawItem.(type) {
+			case string:
+				roles = append(roles, item)
+			case map[string]interface{}:
+				if db, ok := item["db"].(string); ok {
+					if r, ok := item["role"].(string); ok {
+						roles = append(roles, mongodbRole{Role: r, DB: db})
+					}
+				}
+			}
+		}
+	}
+
+	// Store it
+	entry, err := logical.StorageEntryJSON("role/"+name, &roleEntry{
+		DB:    roleDB,
+		MongoDBRoles: roles,
+	})
+	if err != nil {
+		return nil, err
+	}
+	if err := req.Storage.Put(entry); err != nil {
+		return nil, err
+	}
+
+	return nil, nil
+}
+
+type roleEntry struct {
+	DB           string        `json:"db"`
+	MongoDBRoles []interface{} `json:"roles"`
```

Comment:
It's also worth noting that the underlying [`createUser command`](https://docs.mongodb.com/manual/reference/command/createUser/#dbcmd.createUser) used by both the `db.createUser()` shell method and this mongodb secret backend also supports specification of roles in the way I just described. So if a Vault user specifies the following roles array, the array Vault sends to Mongo when creating the user is identical (though, as BSON rather than JSON):

```
[ "readWrite", { "role": "read", "db": "foo" }, { "role": "readWrite", "db": "bar" } ]
```